File: docs/referencia-apk-ocr/PurchaseExtractor.kt
package com.eric.finanzas.utils

import android.util.Log
import java.util.*

class PurchaseExtractor {

    private val spanishMonths = mapOf(
        "ene" to 0, "feb" to 1, "mar" to 2, "abr" to 3,
        "may" to 4, "jun" to 5, "jul" to 6, "ago" to 7,
        "sep" to 8, "oct" to 9, "nov" to 10, "dic" to 11,
        "enero" to 0, "febrero" to 1, "marzo" to 2, "abril" to 3,
        "mayo" to 4, "junio" to 5, "julio" to 6, "agosto" to 7,
        "septiembre" to 8, "octubre" to 9, "noviembre" to 10, "diciembre" to 11
    )

    private val exclusionWords = listOf(
        "lte", "4g", "5g", "3g", "volte", "vo lte", "wifi",
        "batería", "bateria", "señal", "senal", "operador",
        "visa", "mastercard", "débito", "debito", "crédito", "credito",
        "american express", "dinners",
        "id de transacción", "id de transaccion", "id transacción", "id transaccion",
        "comunicarse con la entidad emisora",
        "comunicate con la entidad emisora",
        "obtén recibos más detallados", "obten recibos mas detallados",
        "configurar", "notificaciones", "edge", "gprs", "battery",
        "movistar", "entel", "claro", "wom", "virgin mobile",
        "tarjeta", "terminada en", "cuotas", "cuota",
        "fecha", "hora", "completada", "domingo", "lunes", "martes",
        "miércoles", "miercoles", "jueves", "viernes", "sábado", "sabado"
    )

    fun extract(allLines: List<LineInfo>, allBlocks: List<BlockInfo>): OcrResult {
        val imageHeight = if (allBlocks.isNotEmpty()) allBlocks.maxOf { it.bottom } else 0
        val (monto, montoBlock) = extractMonto(allBlocks)
        val (fecha, _) = extractFechaHora(allLines)
        val comercio = extractComercio(allBlocks, montoBlock, imageHeight)

        Log.d("PurchaseExtractor", "Comercio: $comercio, Monto: $monto, Fecha: $fecha")

        return OcrResult(
            comercio = comercio,
            monto = monto,
            fecha = fecha,
            tipoDetectado = DocumentType.COMPRA
        )
    }

    private fun extractMonto(blocks: List<BlockInfo>): Pair<Double?, BlockInfo?> {
        for (block in blocks) {
            for (line in block.lines) {
                val text = line.text.trim()
                if (text.isEmpty()) continue

                val clpMatch = Regex(
                    """CLP\s*\$?\s*([\d\s,]+(?:[.,]\d{1,2})?)""",
                    RegexOption.IGNORE_CASE
                ).find(text)
                if (clpMatch != null) {
                    val raw = clpMatch.groupValues[1].trim().replace(" ", "")
                    val parsed = limpiarMonto(raw)
                    if (parsed != null && parsed > 0) return Pair(parsed, block)
                }

                val dollarMatch = Regex("""\$\s*([\d\s,]+(?:[.,]\d{1,2})?)""").find(text)
                if (dollarMatch != null) {
                    val raw = dollarMatch.groupValues[1].trim().replace(" ", "")
                    val parsed = limpiarMonto(raw)
                    if (parsed != null && parsed > 0) return Pair(parsed, block)
                }

                val labelMatch = Regex(
                    """(?:Total|Monto|Importe|Valor|Pago)\s*:?\s*\$?\s*([\d\s,]+(?:[.,]\d{1,2})?)""",
                    RegexOption.IGNORE_CASE
                ).find(text)
                if (labelMatch != null) {
                    val raw = labelMatch.groupValues[1].trim().replace(" ", "")
                    val parsed = limpiarMonto(raw)
                    if (parsed != null && parsed > 0) return Pair(parsed, block)
                }
            }
        }
        return Pair(null, null)
    }

    private fun limpiarMonto(raw: String): Double? {
        try {
            return if (raw.contains(",") && raw.contains(".")) {
                val lastCommaIdx = raw.lastIndexOf(",")
                val integerPart = raw.substring(0, lastCommaIdx).replace(".", "")
                val decimalPart = raw.substring(lastCommaIdx + 1).take(2)
                "$integerPart.$decimalPart".toDoubleOrNull()
            } else if (raw.contains(",")) {
                val parts = raw.split(",")
                if (parts.size == 2 && parts[1].length <= 2) {
                    raw.replace(",", ".").toDoubleOrNull()
                } else {
                    raw.replace(",", "").toDoubleOrNull()
                }
            } else if (raw.contains(".")) {
                val parts = raw.split("\\.".toRegex())
                if (parts.size == 2 && parts[1].length <= 2) {
                    raw.toDoubleOrNull()
                } else {
                    raw.replace(".", "").toDoubleOrNull()
                }
            } else {
                raw.toDoubleOrNull()
            }
        } catch (_: Exception) {
            return null
        }
    }

    private fun extractFechaHora(lines: List<LineInfo>): Pair<Long?, String?> {
        val dateRegex = Regex(
            """(\d{1,2})\s+de\s+([a-zA-Záéíóúñ]+)\s+a\s+las\s+(\d{1,2}):(\d{2})"""
        )

        for (line in lines) {
            val match = dateRegex.find(line.text.trim())
            if (match != null) {
                val day = match.groupValues[1].toIntOrNull() ?: continue
                val monthStr = match.groupValues[2].lowercase(Locale.ROOT)
                val month = spanishMonths[monthStr] ?: continue
                val hour = match.groupValues[3].toIntOrNull() ?: continue
                val minute = match.groupValues[4].toIntOrNull() ?: continue

                if (day < 1 || day > 31 || hour > 23 || minute > 59) continue

                val cal = Calendar.getInstance()
                cal.set(Calendar.YEAR, cal.get(Calendar.YEAR))
                cal.set(Calendar.MONTH, month)
                cal.set(Calendar.DAY_OF_MONTH, day)
                cal.set(Calendar.HOUR_OF_DAY, hour)
                cal.set(Calendar.MINUTE, minute)
                cal.set(Calendar.SECOND, 0)
                cal.set(Calendar.MILLISECOND, 0)

                val horaStr = "${hour.toString().padStart(2, '0')}:${minute.toString().padStart(2, '0')}"
                return Pair(cal.timeInMillis, horaStr)
            }
        }

        for (line in lines) {
            val text = line.text.trim()
            if (text.contains("/")) {
                val parts = text.split("/")
                if (parts.size >= 3) {
                    val day = parts[0].filter { it.isDigit() }.toIntOrNull()
                    val month = parts[1].filter { it.isDigit() }.toIntOrNull()
                    var year = parts[2].substring(0, minOf(4, parts[2].length)).filter { it.isDigit() }.toIntOrNull()
                    if (day != null && month != null && year != null && day in 1..31 && month in 1..12) {
                        if (year < 100) year += 2000
                        val timeMatch = Regex("""(\d{1,2}):(\d{2})""").find(text)
                        val hour = timeMatch?.groupValues?.get(1)?.toIntOrNull() ?: 0
                        val minute = timeMatch?.groupValues?.get(2)?.toIntOrNull() ?: 0

                        val cal = Calendar.getInstance()
                        cal.set(Calendar.YEAR, year)
                        cal.set(Calendar.MONTH, month - 1)
                        cal.set(Calendar.DAY_OF_MONTH, day)
                        cal.set(Calendar.HOUR_OF_DAY, hour)
                        cal.set(Calendar.MINUTE, minute)
                        cal.set(Calendar.SECOND, 0)
                        cal.set(Calendar.MILLISECOND, 0)

                        val horaStr = "${hour.toString().padStart(2, '0')}:${minute.toString().padStart(2, '0')}"
                        return Pair(cal.timeInMillis, horaStr)
                    }
                }
            }
        }

        return Pair(null, null)
    }

    private fun extractComercio(
        blocks: List<BlockInfo>,
        montoBlock: BlockInfo?,
        imageHeight: Int
    ): String? {
        if (montoBlock == null) return null

        val statusBarThreshold = (imageHeight * 0.08f).toInt().coerceIn(50, 100)

        val candidates = blocks.filter { block ->
            block.top > statusBarThreshold &&
            block.bottom <= montoBlock.top &&
            block != montoBlock &&
            !contieneTextoExcluido(block) &&
            !esBloqueMonetario(block) &&
            !esBloqueSecundario(block)
        }

        if (candidates.isEmpty()) return null

        val bestBlock = candidates.minWithOrNull(
            compareBy<BlockInfo> { montoBlock.top - it.bottom }
                .thenByDescending { it.lines.maxOf { line -> line.height } }
                .thenByDescending { it.lines.sumOf { line -> line.text.trim().length } }
        )

        if (bestBlock != null) {
            val name = bestBlock.lines
                .joinToString(" ") { it.text.trim() }
                .trim()
            if (name.isNotBlank()) return name
        }

        return null
    }

    private fun contieneTextoExcluido(block: BlockInfo): Boolean {
        val fullText = block.lines.joinToString(" ") { it.text }.lowercase(Locale.ROOT)
        if (exclusionWords.any { fullText.contains(it) }) return true
        val condensed = block.lines.joinToString("") { it.text }
        if (condensed.length <= 4 && condensed.all { it.isDigit() || it.isLetter() || it in "+-" }) return true
        val digitRatio = condensed.count { it.isDigit() }.toFloat() / condensed.length.coerceAtLeast(1)
        if (digitRatio > 0.6f && condensed.length >= 3) return true
        return false
    }

    private fun esBloqueMonetario(block: BlockInfo): Boolean {
        val text = block.lines.joinToString(" ") { it.text }
        if (text.contains("$")) return true
        if (Regex("""\bCLP\b""", RegexOption.IGNORE_CASE).containsMatchIn(text)) return true
        if (Regex("""^(Monto|Total|Importe|Valor|Pago)\b""", RegexOption.IGNORE_CASE).containsMatchIn(text)) return true
        if (Regex("""^\$?\s*[\d]{1,3}(?:[.,]\d{3})*(?:[.,]\d{1,2})?\s*$""").matches(text.trim())) return true
        return false
    }

    private fun esBloqueSecundario(block: BlockInfo): Boolean {
        val text = block.lines.joinToString(" ") { it.text }.lowercase(Locale.ROOT)
        if (text.contains("a las")) return true
        if (text.startsWith("completada")) return true
        if (Regex("""\d{1,2}/\d{1,2}/\d{2,4}""").containsMatchIn(text)) return true
        if (Regex("""^\d{1,2}:\d{2}(:\d{2})?$""").matches(text.trim())) return true
        if (text.contains("gracias por tu compra")) return true
        if (text.contains("aprobada")) return true
        if (text.length <= 3) return true
        return false
    }
}

File: src/ocr/purchaseExtractor.js
// Port de PurchaseExtractor.kt.
//
// Diferencia con el original: las 3 regex de monto (CLP/$/etiqueta) tenían
// un bug en el Kotlin original — el grupo de captura era
// `[\d\s,]+(?:[.,]\d{1,2})?`, que no reconoce el separador de miles con
// punto ("$8.500" se truncaba a "8.50" -> 8,5). Se corrige acá usando el
// mismo patrón de miles que ya funciona en transferenciaParser.js
// (`[\d]+(?:[.,]\d{3})*(?:[.,]\d{1,2})?`), decisión tomada explícitamente
// al portar (ver Fase 4c). El resto es puerto 1:1.
//
// La fecha se devuelve como Date (hora local), no epoch millis.

const MESES_ES = {
  ene: 0, feb: 1, mar: 2, abr: 3, may: 4, jun: 5, jul: 6, ago: 7, sep: 8, oct: 9, nov: 10, dic: 11,
  enero: 0, febrero: 1, marzo: 2, abril: 3, mayo: 4, junio: 5, julio: 6, agosto: 7,
  septiembre: 8, octubre: 9, noviembre: 10, diciembre: 11,
};

const EXCLUSION_WORDS = [
  "lte", "4g", "5g", "3g", "volte", "vo lte", "wifi",
  "batería", "bateria", "señal", "senal", "operador",
  "visa", "mastercard", "débito", "debito", "crédito", "credito",
  "american express", "dinners",
  "id de transacción", "id de transaccion", "id transacción", "id transaccion",
  "comunicarse con la entidad emisora",
  "comunicate con la entidad emisora",
  "obtén recibos más detallados", "obten recibos mas detallados",
  "configurar", "notificaciones", "edge", "gprs", "battery",
  "movistar", "entel", "claro", "wom", "virgin mobile",
  "tarjeta", "terminada en", "cuotas", "cuota",
  "fecha", "hora", "completada", "domingo", "lunes", "martes",
  "miércoles", "miercoles", "jueves", "viernes", "sábado", "sabado",
];

const MONTO_NUM = "[\\d]+(?:[.,]\\d{3})*(?:[.,]\\d{1,2})?";
const CLP_REGEX = new RegExp(`CLP\\s*\\$?\\s*(${MONTO_NUM})`, "i");
const DOLLAR_REGEX = new RegExp(`\\$\\s*(${MONTO_NUM})`);
const LABEL_MONTO_REGEX = new RegExp(`(?:Total|Monto|Importe|Valor|Pago)\\s*:?\\s*\\$?\\s*(${MONTO_NUM})`, "i");
const FECHA_TEXTO_REGEX = /(\d{1,2})\s+de\s+([a-zA-Záéíóúñ]+)\s+a\s+las\s+(\d{1,2}):(\d{2})/;
const HORA_REGEX = /(\d{1,2}):(\d{2})/;

function aEnteroONull(s) {
  if (typeof s !== "string" || !/^\d+$/.test(s.trim())) return null;
  return parseInt(s, 10);
}

function aDecimalONull(s) {
  if (typeof s !== "string" || s.trim() === "") return null;
  const n = Number(s);
  return Number.isFinite(n) ? n : null;
}

function soloDigitos(s) {
  return (s.match(/\d/g) || []).join("");
}

function limpiarMonto(raw) {
  if (raw.includes(",") && raw.includes(".")) {
    const lastCommaIdx = raw.lastIndexOf(",");
    const integerPart = raw.slice(0, lastCommaIdx).replace(/\./g, "");
    const decimalPart = raw.slice(lastCommaIdx + 1).slice(0, 2);
    return aDecimalONull(`${integerPart}.${decimalPart}`);
  }
  if (raw.includes(",")) {
    const parts = raw.split(",");
    if (parts.length === 2 && parts[1].length <= 2) return aDecimalONull(raw.replace(",", "."));
    return aDecimalONull(raw.replace(/,/g, ""));
  }
  if (raw.includes(".")) {
    const parts = raw.split(".");
    if (parts.length === 2 && parts[1].length <= 2) return aDecimalONull(raw);
    return aDecimalONull(raw.replace(/\./g, ""));
  }
  return aDecimalONull(raw);
}

function buscarMontoEnTexto(text) {
  const clpMatch = CLP_REGEX.exec(text);
  if (clpMatch) {
    const parsed = limpiarMonto(clpMatch[1].trim());
    if (parsed !== null && parsed > 0) return parsed;
  }

  const dollarMatch = DOLLAR_REGEX.exec(text);
  if (dollarMatch) {
    const parsed = limpiarMonto(dollarMatch[1].trim());
    if (parsed !== null && parsed > 0) return parsed;
  }

  const labelMatch = LABEL_MONTO_REGEX.exec(text);
  if (labelMatch) {
    const parsed = limpiarMonto(labelMatch[1].trim());
    if (parsed !== null && parsed > 0) return parsed;
  }

  return null;
}

function extractMonto(blocks) {
  for (const block of blocks) {
    for (const line of block.lines) {
      const parsed = buscarMontoEnTexto(line.text.trim());
      if (parsed !== null) return [parsed, block];
    }
  }

  // Fallback: montos grandes (ej. "CLP1,350") a veces quedan en líneas
  // separadas dentro del mismo bloque por la diferencia de tamaño de fuente
  // entre el prefijo de moneda y el número — se reintenta uniendo todo el
  // texto del bloque.
  for (const block of blocks) {
    const parsed = buscarMontoEnTexto(block.lines.map((l) => l.text.trim()).join(" "));
    if (parsed !== null) return [parsed, block];
  }

  return [null, null];
}

function extractFechaHora(lines) {
  for (const line of lines) {
    const match = FECHA_TEXTO_REGEX.exec(line.text.trim());
    if (match) {
      const day = aEnteroONull(match[1]);
      const mesTexto = match[2].toLowerCase();
      // Google Wallet abrevia "septiembre" como "sept" (4 letras); el resto
      // de las abreviaturas chilenas usan 3 — se acepta el prefijo de 3 como
      // fallback en vez de listar cada variante posible.
      const month = MESES_ES[mesTexto] ?? MESES_ES[mesTexto.slice(0, 3)];
      const hour = aEnteroONull(match[3]);
      const minute = aEnteroONull(match[4]);
      if (day === null || month === undefined || hour === null || minute === null) continue;
      if (day < 1 || day > 31 || hour > 23 || minute > 59) continue;
      const year = new Date().getFullYear();
      const horaStr = `${String(hour).padStart(2, "0")}:${String(minute).padStart(2, "0")}`;
      return [new Date(year, month, day, hour, minute, 0, 0), horaStr];
    }
  }

  for (const line of lines) {
    const text = line.text.trim();
    if (!text.includes("/")) continue;
    const parts = text.split("/");
    if (parts.length < 3) continue;
    const day = aEnteroONull(soloDigitos(parts[0]));
    const month = aEnteroONull(soloDigitos(parts[1]));
    let year = aEnteroONull(soloDigitos(parts[2].slice(0, Math.min(4, parts[2].length))));
    if (day === null || month === null || year === null) continue;
    if (day < 1 || day > 31 || month < 1 || month > 12) continue;
    if (year < 100) year += 2000;
    const timeMatch = HORA_REGEX.exec(text);
    const hour = timeMatch ? aEnteroONull(timeMatch[1]) ?? 0 : 0;
    const minute = timeMatch ? aEnteroONull(timeMatch[2]) ?? 0 : 0;
    const horaStr = `${String(hour).padStart(2, "0")}:${String(minute).padStart(2, "0")}`;
    return [new Date(year, month - 1, day, hour, minute, 0, 0), horaStr];
  }

  return [null, null];
}

function contieneTextoExcluido(block) {
  const fullText = block.lines.map((l) => l.text).join(" ").toLowerCase();
  if (EXCLUSION_WORDS.some((w) => fullText.includes(w))) return true;
  const condensed = block.lines.map((l) => l.text).join("");
  if (condensed.length <= 4 && [...condensed].every((c) => /[\p{L}\p{N}+\-]/u.test(c))) return true;
  const digitRatio = (condensed.match(/\d/g) || []).length / Math.max(condensed.length, 1);
  if (digitRatio > 0.6 && condensed.length >= 3) return true;
  return false;
}

function esBloqueMonetario(block) {
  const text = block.lines.map((l) => l.text).join(" ");
  if (text.includes("$")) return true;
  if (/\bCLP\b/i.test(text)) return true;
  if (/^(Monto|Total|Importe|Valor|Pago)\b/i.test(text)) return true;
  if (/^\$?\s*[\d]{1,3}(?:[.,]\d{3})*(?:[.,]\d{1,2})?\s*$/.test(text.trim())) return true;
  return false;
}

function esBloqueSecundario(block) {
  const text = block.lines.map((l) => l.text).join(" ").toLowerCase();
  if (text.includes("a las")) return true;
  if (text.startsWith("completada")) return true;
  if (/\d{1,2}\/\d{1,2}\/\d{2,4}/.test(text)) return true;
  if (/^\d{1,2}:\d{2}(:\d{2})?$/.test(text.trim())) return true;
  if (text.includes("gracias por tu compra")) return true;
  if (text.includes("aprobada")) return true;
  if (text.length <= 3) return true;
  return false;
}

function extractComercio(blocks, montoBlock, imageHeight) {
  if (!montoBlock) return null;
  const statusBarThreshold = Math.min(100, Math.max(50, Math.trunc(imageHeight * 0.08)));

  const candidates = blocks.filter(
    (block) =>
      block.top > statusBarThreshold &&
      block.bottom <= montoBlock.top &&
      block !== montoBlock &&
      !contieneTextoExcluido(block) &&
      !esBloqueMonetario(block) &&
      !esBloqueSecundario(block)
  );
  if (candidates.length === 0) return null;

  const conClave = candidates.map((block) => ({
    block,
    distancia: montoBlock.top - block.bottom,
    alturaMax: Math.max(...block.lines.map((l) => l.height)),
    largoTotal: block.lines.reduce((s, l) => s + l.text.trim().length, 0),
  }));
  conClave.sort(
    (a, b) => a.distancia - b.distancia || b.alturaMax - a.alturaMax || b.largoTotal - a.largoTotal
  );
  const bestBlock = conClave[0].block;

  const name = bestBlock.lines.map((l) => l.text.trim()).join(" ").trim();
  return name || null;
}

// lineas/bloques: ver construirBloques.js. Devuelve
// { comercio, monto, fecha } — nunca null (el documento ya se clasificó
// como compra antes de llamar acá).
export function parsearCompra(lineas, bloques) {
  const imageHeight = bloques.length ? Math.max(...bloques.map((b) => b.bottom)) : 0;
  const [monto, montoBlock] = extractMonto(bloques);
  const [fecha] = extractFechaHora(lineas);
  const comercio = extractComercio(bloques, montoBlock, imageHeight);
  return { comercio, monto, fecha };
}
